### PR TITLE
[dynamo][eval_frame] Set destroy_extra_state deleter as part of co_extra

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107388
* #107358
* __->__ #107117
* #106917

Using the `freefunc` facility to free the ExtraState objects - https://peps.python.org/pep-0523/

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

**Comment by ezyang** [comment]
letting @jansel on this one

File: torch/csrc/dynamo/eval_frame.c [hunk 1]
```diff
@@ -420,7 +425,6 @@ inline static void set_extra_state(PyCodeObject* code, ExtraState* extra_state)
   //  set_extra_state responsibility to clean it up. It will be deleled during
   //  the reset_code/skip, when the set_extra_state is called with
   //  NULL/SKIP_CODE.
-  destroy_extra_state(code);
   _PyCode_SetExtra((PyObject*)code, extra_index, extra_state);
```
Comment (jansel) [inline torch/csrc/dynamo/eval_frame.c:428]: I am assuming `_PyCode_SetExtra` calls `destroy_extra_state` if the value is already set?

(I have not confirmed this.)

**Review by jansel — APPROVED** [review]

File: torch/csrc/dynamo/eval_frame.c (hunk 1 above)
Comment (anijain2305) [reply torch/csrc/dynamo/eval_frame.c:428]: Yes, it does

image

**Pull request closed without merging**